### DISCUSSION TITLE
Multithreading and connection issues？

### DISCUSSION THREAD

**Discussion by muyiwq** [microsoft/msquic#1769] [General]
Dear Mr
System: centos 7 16 core 32G

I have two problems when I test perf and sample?

1. When testing perf, WAN tests about 23000 HPS and more than 2000 HPS.
But there is a phenomenon I don't understand, why when the total number of connections in the RPS test exceeds 800, there will be connection failures. The more connections, the higher the percentage of failures.
E.g:
WARNING: Only 697 (of 800) connections connected successfully.
WARNING: Only 1045 (of 1500) connections connected successfully.
WARNING: Only 1923 (of 3000) connections connected successfully.
WARNING: Only 4375 (of 8000) connections connected successfully.

After the parallel of HPS is greater than 500, the connection will be stuck and the HPS is lower than 1000. The gap between each test of HPS is relatively large, parallel 100-300 is generally around 2000 and occasionally 1000.
I guess there is a certain limit on the number of connections processed by each CPU core?
The number of connections is unstable and the connection fails. How to solve this situation?

2. Multi-threaded operation sample to establish multi-connection single-stream performance test.

The first way:
 <1>MsQuicOpen()<2>RegistrationOpen()<3>ClientLoadConfiguration() and then establish a multi-threaded RunClient()
(20 threads) Each thread establishes multiple connections and a single stream HRPS is about 460.
The number of threads is different, the HRPS is different. .

The second way:
But when I start multithreading outside the entire client (before MsQuicOpen), ConfigurationClose or QuicWorkerThread will report an error. Is this usage wrong?
Screenshot below,

I now want to realize the HRPS multi-connection single stream scenario, how should I improve it?

I very much look forward to your reply. . .
![thread_error](https://user-images.githubusercontent.com/34933482/123373632-44ce4680-d5b8-11eb-816b-f83d29b714fa.jpg)

**Comment by nibanks** [comment]
Can you check to see if this might be causing you problems?
https://github.com/microsoft/msquic/blob/main/docs/TSG.md#linux-file-handle-limit-too-small
